server/auth/sso/sso.go

```diff
@@ -0,0 +1,169 @@
+package sso
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+	"strings"
+	"time"
+
+	"github.com/argoproj/pkg/jwt/zjwt"
+	"github.com/argoproj/pkg/rand"
+	"github.com/coreos/go-oidc"
+	log "github.com/sirupsen/logrus"
+	"golang.org/x/oauth2"
+	apiv1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	corev1 "k8s.io/client-go/kubernetes/typed/core/v1"
+
+	wfv1 "github.com/argoproj/argo/pkg/apis/workflow/v1alpha1"
+)
+
+const Prefix = "Bearer id_token:"
+
+type claims struct {
+	Groups []string `json:"groups"`
+}
+
+type Interface interface {
+	Authorize(ctx context.Context, authorization string) (wfv1.User, error)
+	HandleRedirect(writer http.ResponseWriter, request *http.Request)
+	HandleCallback(writer http.ResponseWriter, request *http.Request)
+}
+
+type sso struct {
+	config          *oauth2.Config
+	idTokenVerifier *oidc.IDTokenVerifier
+	baseHRef        string
+	secure          bool
+}
+
+type Config struct {
+	Issuer       string                  `json:"issuer"`
+	ClientID     string                  `json:"clientId"`
+	ClientSecret apiv1.SecretKeySelector `json:"clientSecret"`
+	RedirectURL  string                  `json:"redirectUrl"`
+}
+
+func New(c Config, secretsIf corev1.SecretInterface, baseHRef string, secure bool) (Interface, error) {
+	if c.Issuer == "" {
+		return nil, fmt.Errorf("issuer empty")
+	}
+	if c.ClientID == "" {
+		return nil, fmt.Errorf("clientId empty")
+	}
+	if c.ClientSecret.Name == "" || c.ClientSecret.Key == "" {
+		return nil, fmt.Errorf("clientSecret empty")
+	}
+	if c.RedirectURL == "" {
+		return nil, fmt.Errorf("redirectUrl empty")
+	}
+	secrets, err := secretsIf.Get(c.ClientSecret.Name, metav1.GetOptions{})
+	if err != nil {
+		return nil, err
+	}
+	provider, err := oidc.NewProvider(context.Background(), c.Issuer)
+	if err != nil {
+		return nil, err
+	}
+	config := &oauth2.Config{
+		ClientID:     c.ClientID,
+		ClientSecret: string(secrets.Data[c.ClientSecret.Key]),
+		RedirectURL:  c.RedirectURL,
+		Endpoint:     provider.Endpoint(),
+		Scopes:       []string{oidc.ScopeOpenID, "groups"},
+	}
+	idTokenVerifier := provider.Verifier(&oidc.Config{ClientID: config.ClientID})
+	log.WithFields(log.Fields{"redirectUrl": config.RedirectURL, "issuer": c.Issuer, "clientId": c.ClientID}).Info("SSO configuration")
+	return &sso{config, idTokenVerifier, baseHRef, secure}, nil
+}
+
+const stateCookieName = "oauthState"
+
+func (s *sso) HandleRedirect(w http.ResponseWriter, r *http.Request) {
+	state := rand.RandString(10)
+	http.SetCookie(w, &http.Cookie{
+		Name:     stateCookieName,
+		Value:    state,
+		Expires:  time.Now().Add(1 * time.Hour),
```

Comment:
I believe state nonces are supposed to redeem on the order of minutes. I would set this much lower: 3 minutes.